File: js/075_SortColors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    var r = -1, w = -1, b = -1;
    for(var i in nums){
        if(nums[i] === 0){
            nums[++b] = 2;
            nums[++w] = 1;
            nums[++r] = 0;
        }else if(nums[i] === 1){
            nums[++b] = 2;
            nums[++w] = 1;
        }else if(nums[i] === 2){
            nums[++b] = 2;
        }
    }
};

//swap maybe not one pass
var sortColors = function(nums) {
    var red = 0, blue = nums.length-1;
    for(var i = 0; i <= blue; ++i){
        while(nums[i] === 2 && i < blue){
            var t = nums[i];
            nums[i] = nums[blue];
            nums[blue] = t;
            blue--;
        }
        while(nums[i] === 0 && i > red){
            var h = nums[i];
            nums[i] = nums[red];
            nums[red] = h;
            red++;
        }
    }
};

File: cpp/075_SortColors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int red = -1, white = -1, blue = -1;
        for(auto num : nums){
            if(num == 0){
                nums[++blue] = 2;
                nums[++white] = 1;
                nums[++red] = 0;
            }else if(num == 1){
                nums[++blue] = 2;
                nums[++white] = 1;
            }else if(num == 2){
                nums[++blue] = 2;
            }
        }
    }
};

//swap maybe not one pass
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int red = 0, blue = nums.size()-1;
        for(int i = 0; i <= blue; ++i){
            while(nums[i] == 2 && i < blue) swap(nums[i],nums[blue--]);
            while(nums[i] == 0 && i > red) swap(nums[i],nums[red++]);
        }
    }
private:
    void swap(int& x, int& y){
        int t = x;
        x = y;
        y = t;
    }
};